### PR TITLE
Continuous batching 

### PR DESCRIPTION
**Pull request by quic-rishinr**
- Added support for transforming, compiling, and executing models using continuous batching (CB) mode.
- Supported architectures for CB execution include Llama, Mistral, Mixtral, and Starcoder.
- Integrated continuous batching transform under PyTorch transform.
- Added support for setting QEFF_HOME and XDG_CACHE_HOME directory paths.
- Refactored text_generation_inference module to accommodate CB mode execution and enable switching between regular and CB execution.
- Updated README and documentation to reflect the new changes.
- Added tests for continuous batching mode.

### CODE REVIEW

File: QEfficient/utils/constants.py [hunk 1]
```diff
@@ -10,7 +10,7 @@
 UTILS_DIR = os.path.dirname(os.path.abspath(__file__))
 QEFF_DIR = os.path.dirname(UTILS_DIR)
 ROOT_DIR = os.path.dirname(QEFF_DIR)
-QEFF_MODELS_DIR = os.path.join(ROOT_DIR, "qeff_models")
+QEFF_MODELS_DIR = os.environ.get("QEFF_MODELS_DIR", os.path.join(ROOT_DIR, "qeff_models"))
```
Comment (vbaddi) [inline QEfficient/utils/constants.py:13]: You can pick this change from `release/v1.16` branch

**Review by vbaddi — CHANGES_REQUESTED** [review]

File: QEfficient/utils/constants.py (hunk 1 above)
Comment (quic-rishinr) [reply QEfficient/utils/constants.py]: I have added the changes in the recent commit.

File: QEfficient/exporter/export_hf_to_cloud_ai_100.py [hunk 2]
```diff
@@ -242,6 +249,10 @@ def export_kvstyle_transformed_model_to_onnx(
         for _ in range(n_layer)
     ]
 
+    model_architecture = config.architectures[0]
+    if full_batch_size and model_architecture not in get_lists_of_cb_qeff_models.architectures:
+        raise Exception(f"Continuous batching support for {model_architecture} is not enabled currently")
```
Comment (ochougul) [inline QEfficient/exporter/export_hf_to_cloud_ai_100.py:254]: raise `NotImplementedError` ?

Comment (ochougul) [reply QEfficient/exporter/export_hf_to_cloud_ai_100.py:254]: Can we check this at the beginning of `export` function? so user doesn't need to wait till this point?

File: QEfficient/cloud/export.py [hunk 3]
```diff
@@ -71,8 +77,13 @@ def main(
     :hf_token: str. HuggingFace login token to access private repos.
     :local_model_dir: str. Path to custom model weights and config files.
     """
-    cache_dir = check_and_assign_cache_dir(local_model_dir, cache_dir)
-    get_onnx_model_path(model_name=model_name, cache_dir=cache_dir, hf_token=hf_token, local_model_dir=local_model_dir)
+    get_onnx_model_path(
```
Comment (ochougul) [inline QEfficient/cloud/export.py:80]: why did you remove `check_and_assign_cache_dir` ?

File: QEfficient/cloud/export.py [hunk 4]
```diff
@@ -25,11 +25,13 @@ def get_onnx_model_path(
     tokenizer: Optional[Union[PreTrainedTokenizerFast, PreTrainedTokenizer]] = None,
     hf_token: Optional[str] = None,
     local_model_dir: Optional[str] = None,
+    full_batch_size: Optional[int] = None,
+    base_dir_name: Optional[str] = "",
```
Comment (ochougul) [inline QEfficient/cloud/export.py:29]: `base_dir_name` not added to parser arguments?

File: QEfficient/cloud/infer.py [hunk 5]
```diff
@@ -26,6 +26,8 @@
 4. Download HF model -> transform -> export -> compile -> execute
 """
 
+# TODO update the heck_batch_size_and_num_prompts method for FBS
+
```
Comment (ochougul) [inline QEfficient/cloud/infer.py:30]: ?

File: QEfficient/customop/ctx_scatter_gather.py [hunk 6]
```diff
@@ -79,3 +79,84 @@ def setup_context(ctx, inputs, outputs):
     @staticmethod
     def symbolic(g: torch.Graph, data: torch.Value, ctx_indices: torch.Value) -> torch.Value:
         return g.onnxscript_op(CtxGather, data, ctx_indices).setTypeAs(data)
+
+
+@onnxscript.script(onnxscript.values.Opset("com.qualcomm.cloud", 1))
```
Comment (ochougul) [inline QEfficient/customop/ctx_scatter_gather.py:84]: remove?

File: QEfficient/exporter/export_utils.py [hunk 7]
```diff
@@ -62,10 +63,14 @@ def export_onnx(
...
             # KV-cache (batch_size, num_heads, past_len, embed_dim)
-            dynamic_axes[iname] = {0: "batch_size", 2: "ctx_len"}
+            dynamic_axes[iname] = {0: dynamic_axis_past_key, 2: "ctx_len"}
+        elif iname == "batch_index":
+            dynamic_axes[iname] = {0: "batch_size"}
+
     if "past_key.0" in input_names and "attention_mask" in input_names:
         dynamic_axes["attention_mask"] = {0: "batch_size", 1: "ctx_len"}
 
+    custom_opsets = {"com.qti.aisw.onnx": 1} if dynamic_axis_past_key == "full_batch_size" else None
```
Comment (ochougul) [inline QEfficient/exporter/export_utils.py:73]: Is this required?

File: QEfficient/utils/_utils.py [hunk 8]
```diff
@@ -93,7 +93,18 @@ def qpc_exists(qpc_dir_path: str) -> bool:
...
 
 
-def onnx_exists(model_name: str) -> Tuple[bool, str, str]:
+def get_onnx_dir_name(model_name, has_fbs):
+    # Create a unique directory name for the ONNX model
+    # Clearly indicate whether it's with or without FBS
+    # Replace all hyphens with underscores
+    model_name_safe = model_name.replace("/", "_").replace("-", "_")
+    if has_fbs:
+        return f"onnx_{model_name_safe}_with_fbs"
```
Comment (ochougul) [inline QEfficient/utils/_utils.py:102]: should this also save info about value of fps in the name?

File: QEfficient/utils/constants.py [hunk 9]
```diff
@@ -10,7 +10,37 @@
...
+    Determine the directory for storing QEFF models.
+    Priority:
+    1. Use $XDG_CACHE_HOME/qeff_models if XDG_CACHE_HOME is set.
+    2. Use QEFF_HOME if set in environment.
+    3. Default to ~/.cache/qeff_models.
+    Sets QEFF_MODELS_DIR environment variable if not already set.
+    Returns:
+        str: Path to the QEFF models directory.
+    """
+    qeff_cache_home = os.environ.get("QEFF_HOME")
```
Comment (ochougul) [inline QEfficient/utils/constants.py:27]: Need to add `QEFF_HOME` throughout the repo?

**Review by ochougul — CHANGES_REQUESTED** [review]
Please rebase

File: QEfficient/cloud/export.py (hunk 4 above)
Comment (quic-rishinr) [reply QEfficient/cloud/export.py:29]: No its not used now. Will remove it

File: QEfficient/cloud/infer.py (hunk 5 above)
Comment (quic-rishinr) [reply QEfficient/cloud/infer.py:27]: removed it.

File: QEfficient/customop/ctx_scatter_gather.py (hunk 6 above)
Comment (quic-rishinr) [reply QEfficient/customop/ctx_scatter_gather.py:84]: Didnt get you. Additional space? Will remove CB specific CtxGatherFunc and CtxScatterFunc from stx_scatter_gather.py as its maintained in different file.

File: QEfficient/utils/_utils.py (hunk 8 above)
Comment (quic-rishinr) [reply QEfficient/utils/_utils.py:102]: No. Model should be same for any value of FBS. Its a dynamic variable input that we can provide through specialization file.

File: QEfficient/exporter/export_hf_to_cloud_ai_100.py (hunk 2 above)
Comment (quic-rishinr) [reply QEfficient/exporter/export_hf_to_cloud_ai_100.py:265]: Changed it to NotImplementedError  and moved it under export_for_cloud

File: QEfficient/exporter/export_utils.py (hunk 7 above)
Comment (quic-rishinr) [reply QEfficient/exporter/export_utils.py:75]: This can be removed. Will remove it from the the PR

File: QEfficient/cloud/export.py (hunk 3 above)
Comment (quic-rishinr) [reply QEfficient/cloud/export.py:81]: Had picked up this change from Release 1.16 branch https://github.com/quic/efficient-transformers/pull/69/files

File: QEfficient/utils/constants.py (hunk 9 above)
Comment (quic-rishinr) [reply QEfficient/utils/constants.py:27]: had picked up this from Release 1.16 https://github.com/quic/efficient-transformers/pull/69/files

File: QEfficient/exporter/export_hf_to_cloud_ai_100.py [hunk 10]
```diff
@@ -206,6 +208,57 @@ def export_kvstyle_transformed_model_to_onnx(
...
+    batch_size, prompt_len = inputs["input_ids"].shape
+    # inputs.pop("attention_mask")
+    inputs["position_ids"] = torch.arange(prompt_len).view(1, -1)
+    if full_batch_size:
+        inputs["position_ids"] = torch.arange(prompt_len).view(1, prompt_len)  # FIXME remove is this is not required
+        batch_index = torch.arange(1).view(-1, 1)
+        inputs["batch_index"] = batch_index
+
+    config = transformed_model.config
+    if hasattr(config, "n_head"):  # Assuming n_head is a key in the config (GPTs/CodeGen)
```
[the remark this answers is not in the document]
Comment (quic-rishinr) [reply QEfficient/exporter/export_hf_to_cloud_ai_100.py]: Adhered to the InputHandler method for fetching pytorch inputs

File: QEfficient/cloud/execute.py [hunk 11]
```diff
@@ -90,6 +92,9 @@ def main(
         required=False,
         help="Cache dir to store HF Downloads",
     )
+    parser.add_argument(
```
[the remark this answers is not in the document]
Comment (quic-rishinr) [reply QEfficient/cloud/execute.py]: Added support for fetching FBS from session_bindings during execution

File: QEfficient/exporter/export_hf_to_cloud_ai_100.py [hunk 12]
```diff
@@ -216,11 +269,31 @@ def export_kvstyle_transformed_model_to_onnx(
...
+        inputs["position_ids"][batch_index.view(-1)] = position_ids
+
+        inputs["batch_index"] = torch.arange(full_batch_size).view(-1, 1)
+    else:
+        inputs = input_handler.update_pytorch_inputs(inputs, pt_outputs)
+        # To avoid issues in onnx export
+        inputs["position_ids"] = torch.full((1, 1), seq_len - 1)
 
+    # Run PyTorch inference for decode in loop
+    # todo: vbaddi, fix it to verify on Cloud AI 100.
```
[the remark this answers is not in the document]
Comment (quic-rishinr) [reply QEfficient/exporter/export_hf_to_cloud_ai_100.py]: Removed.

File: QEfficient/exporter/export_hf_to_cloud_ai_100.py [hunk 13]
```diff
@@ -216,11 +269,31 @@ def export_kvstyle_transformed_model_to_onnx(
...
+    for i in range(1):
+        pt_outputs = transformed_model(**inputs)
+        inputs["input_ids"] = pt_outputs.logits.detach().argmax(2)
+        inputs["position_ids"] += 1
+
+    # To avoid issues in onnx export
+    inputs["position_ids"] = torch.full((full_batch_size if full_batch_size else batch_size, 1), seq_len - 1)
```
[the remark this answers is not in the document]
Comment (quic-rishinr) [reply QEfficient/exporter/export_hf_to_cloud_ai_100.py]: Removed

File: QEfficient/exporter/export_hf_to_cloud_ai_100.py [hunk 14]
```diff
@@ -216,11 +269,31 @@ def export_kvstyle_transformed_model_to_onnx(
 
     # Build inputs for next iteration from outputs
     # Build inputs for decode
-    inputs = input_handler.update_pytorch_inputs(inputs, pt_outputs)
 
-    # To avoid issues in onnx export
-    inputs["position_ids"] = torch.full((1, 1), seq_len - 1)
+    if full_batch_size:
```
[the remark this answers is not in the document]
Comment (quic-rishinr) [reply QEfficient/exporter/export_hf_to_cloud_ai_100.py]: handled in latest patch

File: QEfficient/exporter/export_hf_to_cloud_ai_100.py [hunk 15]
```diff
@@ -206,6 +208,57 @@ def export_kvstyle_transformed_model_to_onnx(
         ctx_len=seq_len,
     )
     inputs = input_handler.prepare_pytorch_inputs()
+    assert seq_len > 0, "Need seq_len to be greater than zero"
```
[the remark this answers is not in the document]
Comment (quic-rishinr) [reply QEfficient/exporter/export_hf_to_cloud_ai_100.py]: its already handled in the code

File: QEfficient/exporter/export_utils.py [hunk 16]
```diff
@@ -80,7 +85,7 @@ def export_onnx(
             output_names=output_names,
             dynamic_axes=dynamic_axes,
             opset_version=13,
-            custom_opsets={"com.qti.aisw.onnx": 1},
+            # custom_opsets=custom_opsets,
```
[the remark this answers is not in the document]
Comment (quic-rishinr) [reply QEfficient/exporter/export_utils.py:88]: Updated in the latest patch

File: QEfficient/generation/text_generation_inference.py [hunk 17]
```diff
@@ -167,6 +168,12 @@ def fix_prompts(prompt: List[str], batch_size: int):
                 "Number of prompts are not multiple of batch size, dropping last incomplete batch from given input prompts"
             )
             prompt = prompt[: batch_size * (len(prompt) // batch_size)]
+
+        if full_batch_size:
+            assert batch_size == 1, "Only either batch_size or full_batch_size should be greater than one"
```
[the remark this answers is not in the document]
Comment (quic-rishinr) [reply QEfficient/generation/text_generation_inference.py:173]: Updated to value error

File: QEfficient/generation/text_generation_inference.py [hunk 18]
```diff
@@ -320,44 +209,48 @@ def cloud_ai_100_exec_kv(
...
+        ]
+        prefill_time = np.average([info.prefill_time for info in exec_info])
+        decode_perf = np.average([info.decode_perf for info in exec_info])
+        total_perf = np.average([info.total_perf for info in exec_info])
+        total_time = np.average([info.total_time for info in exec_info])
+        generated_texts = [info.generated_texts for info in exec_info]
+        generated_ids = [info.generated_ids for info in exec_info]
+    else:
+        exec_info = generate_text.cloud_ai_100_exec_kv_helper(prompt=prompt, generation_len=generation_len)
+        # prefill_time = np.average(exec_info.prefill_time)
```
[the remark this answers is not in the document]
Comment (quic-rishinr) [reply QEfficient/generation/text_generation_inference.py]: We dont need to average in CB because we are processing all inputs in one go. In regular execution if the input length exceeds the BS, we loop through the inputs. But in the case of CB the latency we capture is for the entire set of inputs.

File: QEfficient/generation/text_generation_inference.py [hunk 19]
```diff
@@ -167,6 +168,16 @@ def fix_prompts(prompt: List[str], batch_size: int):
                 "Number of prompts are not multiple of batch size, dropping last incomplete batch from given input prompts"
             )
             prompt = prompt[: batch_size * (len(prompt) // batch_size)]
+
+        if full_batch_size:
```
Comment (ochougul) [inline QEfficient/generation/text_generation_inference.py:172]: This is not the right place for checking `full_batch_size` value, we should check it outside of this function.
And  can we have a separate function for `fixing_prompts_for_cb`  -> The logic for this function is same when `len(prompt)< full_batch_size`, whereas in else case, we don't need to truncate the prompts as cb can support it.

File: QEfficient/generation/text_generation_inference.py [hunk 20]
```diff
@@ -320,55 +213,524 @@ def cloud_ai_100_exec_kv(
...
+        total_num_decoded_tokens, prefill_perf, decode_perf, total_perf
+        """
+        prefill_time = loop_start - start + decode_pause_time
+        decode_perf = (total_decoded_tokens) / (end - loop_start - decode_pause_time)
+        total_perf = (total_decoded_tokens) / (end - start)
+        total_time = end - start
+
+        return prefill_time, decode_perf, total_perf, total_time
+
+    def get_compilation_batch_size(self, qpc_path: str):
```
Comment (quic-mamta) [inline QEfficient/generation/text_generation_inference.py:328]: this function is unused.

File: QEfficient/generation/text_generation_inference.py [hunk 21]
```diff
@@ -320,55 +213,524 @@ def cloud_ai_100_exec_kv(
...
+        """
+        decode_inputs = {}
+        decode_inputs["input_ids"] = self.decode_input_ids
+        decode_inputs["position_ids"] = self.decode_pos_ids
+        if self.batch_index is not None:
+            decode_inputs["batch_index"] = self.batch_index
+
+        return decode_inputs
+
+    def prepare_prompt(self, prompt, batch_size):
```
Comment (quic-mamta) [inline QEfficient/generation/text_generation_inference.py:427]: this preparation is already handled in fix_prompts function.

**Review by quic-mamta — CHANGES_REQUESTED** [review]

File: QEfficient/generation/text_generation_inference.py (hunk 19 above)
Comment (ochougul) [reply QEfficient/generation/text_generation_inference.py:184]: we don't need to drop the prompts, in case of full_batch_size can we change the implementation to ->
```
def fix_prompts(prompt: List[str], batch_size: int, full_batch_size: int = None):
    if len(prompt) < batch_size or (full_batch_size is not None and len(prompt) < full_batch_size):
        logger.warning("Number of prompts are less than batch size/full batch size, repeating to required batch size")
        prompt = prompt * -(batch_size // -len(prompt))  # Repeat prompt to required size
        prompt = prompt[:batch_size]  # Truncate prompts to required size
    elif not full_batch_size:
        if (len(prompt) % batch_size) > 0:
            logger.warning(
                "Number of prompts are not multiple of batch size, dropping last incomplete batch from given input prompts"
            )
            prompt = prompt[: batch_size * (len(prompt) // batch_size)]

    return prompt
```
and test the condition of `batch_size=1` when `fbs` is not None outside this function? inside `cloud_ai_100_exec_kv`?

**Review by ochougul — CHANGES_REQUESTED** [review]

**Review by ochougul — CHANGES_REQUESTED** [review]
I also see few commented code lines, please remove them

File: tests/cloud/high_level_testing.json [hunk 22]
```diff
@@ -1,6 +1,6 @@
 {
     "license": "SEE LICENSE IN LICENSE FILE",
-    "model_name" : ["gpt2"],
+    "model_name" : ["gpt2","TinyLlama/TinyLlama-1.1B-Chat-v1.0","Salesforce/codegen-350M-mono","wtang06/mpt-125m-c4"],
```
Comment (ochougul) [inline tests/cloud/high_level_testing.json:3]: Why are you adding all of them back?

File: tests/config.json
Comment (ochougul) [inline tests/config.json:1]: Remove, not used anywhere

Comment (abukhoy) [reply tests/config.json:1]: Removed.

File: tests/cloud/high_level_testing.json (hunk 22 above)
Comment (abukhoy) [reply tests/cloud/high_level_testing.json]: added only one model for CB checking.

**Review by ochougul — APPROVED** [review]
